### DISCUSSION TITLE
Form Fields Exhibit Inconsistent Behavior When Populated

### DISCUSSION THREAD

**Discussion by luzzodev** [fastapi/fastapi#12536] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from fastapi import FastAPI, Form, Depends


router = FastAPI()


class TestForm:
    def __init__(
            self,
            *,
            field_1: Annotated[
                str | None,
                Form(),
            ] = None,
            field_2: Annotated[
                str | None,
                Form(),
            ] = "not_none_not_empty_string"
    ):

        self.field_1 = field_1
        self.field_2 = field_2


@router.post("/test-form")
def do_empty_form_test(form_data: TestForm = Depends()):
    return {
        "field_1": form_data.field_1,
        "field_2": form_data.field_2
    }
```

### Description

I was trying to work on https://github.com/fastapi/fastapi/issues/12240#issue-2539556953 that belongs from https://github.com/fastapi/fastapi/discussions/12182 and i have founded the source of the issue that affect Form in general and not only related to OAuth2PasswordRequestForm.

This is the changed that has created the issue
https://github.com/fastapi/fastapi/commit/7bad7c09757f8a06cf62cc0838082a766065883e#diff-aef3dac481b68359f4edd6974fa3a047cfde595254a4567a560cebc9ccb0673fR790

What happen?

- fiield_1 has a default value: None
- field_2 has a default value: "not_none_not_empty_string"

Test 1: do not send form fields

Request
```
curl --location 'http://0.0.0.0:5000/test-form' 
```

Response 
```
{
    "field_1": null,
    "field_2": "not_none_not_empty_string"
}
```

Test 2: send both form fields but with an empty string

Request
```
curl --location 'http://0.0.0.0:5000/test-form' \
--form 'field_1=""' \
--form 'field_2=""'
```

Response
```
{
    "field_1": "",  <--- Here i expect: null
    "field_2": "not_none_not_empty_string"
}
```

As you can see

-  In Test 1 all is fine, form fields are not set and each field is using his default.
- In Test 2 i was expecting the same response of Test 1 but instead: the input value is used field_1 and default value for field_2.

This example shows how there is an inconsistent behavior in picking or not the default value.
I have personally tested on version 0.115.3 but all version > 0.113.0 should be affected.

Finally, I would like to talk about the auto conversion of empty string to None in Form fields that even if related if think is too much for this topic.

Thanks to all the maintainers and creators for the beautifull work they've done."

### Operating System

Linux, macOS

### Operating System Details

MacOS Sonoma 14.6
Ubuntu Server 22.04

### FastAPI Version

0.115.3

### Pydantic Version

2.9.0

### Python Version

3.11.9

**Comment by YuriiMotov** [comment]
Can approve that behavior is inconsistent.

When you send empty strings it results to `""` (the value that was sent) for `field1` and to `not_none_not_empty_string` (default value) for `field2`.
I expected that since empty string is valid value it should result to empty strings in both cases.

**Reply by luzzodev** [reply]
Thank you for confirming the abnormal behavior.

> I expected that since empty string is valid value it should result to empty strings in both cases.

In the current implementation, FastAPI's intention is to convert all 'str form fields' to their default value, whether they are None or an empty string. 
You can check at [Fastapi.dependencies.utils](https://github.com/fastapi/fastapi/blob/0.115.3/fastapi/dependencies/utils.py)
```
def _get_multidict_value(
    field: ModelField, values: Mapping[str, Any], alias: Union[str, None] = None
) -> Any:
    alias = alias or field.alias
    if is_sequence_field(field) and isinstance(values, (ImmutableMultiDict, Headers)):
        value = values.getlist(alias)
    else:
        value = values.get(alias, None)
    if (
        value is None
        or (
            isinstance(field.field_info, params.Form)
            and isinstance(value, str)  # For type checks
            and value == ""
        )
        or (is_sequence_field(field) and len(value) == 0)
    ):
        if field.required:
            return
        else:
            return deepcopy(field.default)
    return value
```

While I agree with you, I think the topic of converting empty strings to the default in str form fields could be the subject of a separate discussion.

What I am pointing out here is that if an empty string is given as input in a form field and the default is None, then the original value is used. Otherwise, the default value is applied, which is clearly unexpected behavior.

**Comment by YuriiMotov** [comment, marked as answer]
The issue was fixed by https://github.com/fastapi/fastapi/pull/13537 and available in FastAPI 0.123.2 🎉

With the code from initial post it returns
```
{
    "field_1": null,
    "field_2": "not_none_not_empty_string"
}
```

when empty string are passed
```
curl --location 'http://0.0.0.0:5000/test-form' \
--form 'field_1=""' \
--form 'field_2=""'
```
